### ISSUE TITLE
Adding external tests to a collection

### ISSUE THREAD

**Issue by pytestbot** [pytest-dev/pytest#421]
Originally reported by: **Trevor Bekolay (BitBucket: [tbekolay](http://bitbucket.org/tbekolay), GitHub: [tbekolay](http://github.com/tbekolay))**

---

It would be nice if it were possible to run tests from outside of the current package when running py.test on the current package. A possible use case for this functionality would be to have a common set of tests for many packages implementing a certain protocol or standard. Another use case is for testing pluggable features (this was the the origin of the idea; see [this SO question](http://stackoverflow.com/questions/21030987/running-pytest-tests-in-another-python-package)).

Two things are necessary for this to happen. First, you must get a set of tests contained in another package. Second, you must add that set of tests to the current package.

For getting access to external tests, you can do something similar to what occurs when doing `py.test --collect-only` and providing the path to the external package. However, it should additionally be possible to disable all reporting, and to get programmatic access to the tests collected by that run.

For adding external tests to the current collection, one possibility is `pytest_collection_modifyitems`. If the tests can be collected programmatically they can just be added to the items list. However, a more explicit hook like `pytest_collection_addexternalitems` or something might be preferable.

Since I'd like this for one of my projects, I'm motivated to implement this, but it'd be best to discuss a nice API for doing this before starting an implementation.

---
- Bitbucket: https://bitbucket.org/pytest-dev/pytest/issue/421

**Comment by mloning** [comment]
Thanks @nicoddemus - this is really helpful! I'll get back to you in case we have more questions or get stuck, but I think I know what we have to do now.

**Comment by mloning** [comment]
@nicoddemus one follow-up question, is there an easy way to parametrize the fixtures when define as above in the method? Instead of having a single `ForecasterAlgo()` we'd like to iterate over multiple ones. We're currently using `pytest_generate_tests` to do that, but having this in the class as a method seems to the more elegant solution.

**Comment by kalekundert** [comment]
Not sure if that snippet is meant to be pseudo-code, but it's pretty much the right syntax.  The argument just needs to be named `request`:
```python
class TestForecaster(BaseConformanceSuite):

    @pytest.fixture(params=all_forecasters)
    def target(self, request):
        return request.param
```
More info on parametrized fixtures: https://docs.pytest.org/en/6.2.x/fixture.html#fixture-parametrize

**Comment by mloning** [comment]
Thanks @kalekundert

**Comment by KudryashovIlya** [comment]
Hi all!
I'm trying to understand whether it can be implemented or not: I need to add several "external" tests to the current session.
In our project we have 8 environments and the test structure looks like this:

- tests
- 

In test_*.py we declare tests that inherit from common_case (since the cases and steps are repeated and differ only in some steps/conditions).
So, when I change one common case, I need to check how it will work in all environments and in the current implementation I have to do this manually, so the idea arose to automate this.
I tried to deal with `_pytest.python.pytest_pycollect_makeitem`, having previously created a Module instance, but in the end it doesn't work:
```
>>> path = pathlib.Path(<path_to_test)
>>> module = _pytest.python.pytest_collect_file(path, items[0].parent) # items
>>> _pytest.python.pytest_pycollect_makeitem(module, 'test', module.obj)
>>>
```
I tried to do it as suggested [here](https://github.com/pytest-dev/pytest/issues/421#issuecomment-943386533), but either I didn’t understand the idea or it didn’t suit me.
